src/core/stdc/stdarg.d

```diff
@@ -129,25 +158,107 @@ else version (Windows) // Win64
         ap += size_t.sizeof;
         return arg;
     }
+    else
+    {
+        T a;
+        va_arg(ap, a);
+        return a;
+    }
+}
+
 
-    /************
-     * Retrieve and return the next value that is type T.
-     * This is the preferred version.
-     */
-    void va_arg(T)(ref va_list ap, ref T parmn)
+/**
+ * Retrieve and store in parmn the next value that is of type T.
+ */
+void va_arg(T)(ref va_list ap, ref T parmn)
+{
+    version (X86)
+    {
+        parmn = *cast(T*) ap;
+        ap += T.sizeof.alignUp;
+    }
+    else version (Win64)
     {
         static if (T.sizeof > size_t.sizeof)
             parmn = **cast(T**) ap;
         else
             parmn = *cast(T*) ap;
         ap += size_t.sizeof;
     }
+    else version (SysV_x64)
+    {
+        core.internal.vararg.sysv_x64.va_arg!T(ap, parmn);
+    }
+    else version (AAPCS32)
+    {
+        // AAPCS32 section 6.5 B.5: type with alignment >= 8 is 8-byte aligned
+        // instead of normal 4-byte alignment (APCS doesn't do this).
+        if (T.alignof >= 8)
+            ap.__ap = ap.__ap.alignUp!8;
+        auto p = ap.__ap; // TODO: big-endian adjustment?
+        parmn = *cast(T*) p;
+        ap.__ap += T.sizeof.alignUp;
+    }
+    else version (AAPCS64)
+    {
+        static assert(0, "Unsupported platform");
+    }
+    else version (ARM_Any)
+    {
+        auto p = ap; // TODO: big-endian adjustment?
+        parmn = *cast(T*) p;
+        ap += T.sizeof.alignUp;
+    }
+    else version (PPC_Any)
+    {
+        /*
+         * The rules are described in the 64bit PowerPC ELF ABI Supplement 1.9,
+         * available here:
+         * http://refspecs.linuxfoundation.org/ELF/ppc64/PPC-elf64abi-1.9.html#PARAM-PASS
+         */
+
+        // Chapter 3.1.4 and 3.2.3: alignment may require the va_list pointer to first
+        // be aligned before accessing a value
+        if (T.alignof >= 8)
+            ap = ap.alignUp!8;
+        auto p = ap;
+        version (BigEndian)
+            static if (T.sizeof < size_t.sizeof)
+                p += size_t.sizeof - T.sizeof;
+        parmn = *cast(T*) p;
+        ap += T.sizeof.alignUp;
+    }
+    else version (MIPS_Any)
+    {
+        auto p = ap;
+        version (BigEndian)
+            static if (T.sizeof < size_t.sizeof)
+                p += size_t.sizeof - T.sizeof;
+        parmn = *cast(T*) p;
+        ap += T.sizeof.alignUp;
+    }
+    else
+        static assert(0, "Unsupported platform");
```

Comment:
FYI, the 'default' va_arg is that arguments are passed on the stack for any target that doesn't have its own specific abi.

Reply:
Would you prefer such a 'dangerous' default implementation over forcing the guys to look up the ABI docs when porting a new arch?

Reply:
A big warning sign might suffice. Out of scope for this pr though, but the main primary targets should be trivial to add.

Note, you have no static assert for the declaration of va_list type.